src/index.ts

```diff
@@ -1,117 +1,134 @@
-    remarkPlugins,
-  }).process(mdx)) as {contents: string}
+  const {data: frontmatter} = matter(mdxSource)
 
   const dir = path.join(process.cwd(), `__mdx_bundler_fake_dir__`)
-  const entryPath = path.join(dir, './index.mdx.js')
-  const absoluteFiles: Record<string, string> = {
-    [entryPath]: entryCode,
-  }
+  const entryPath = path.join(dir, './index.mdx')
+
+  const absoluteFiles: Record<string, string> = {[entryPath]: mdxSource}
+
   for (const [filepath, fileCode] of Object.entries(files)) {
     absoluteFiles[path.join(dir, filepath)] = fileCode
   }
-  const inMemoryModulePlugin = {
-    name: 'in-memory-module',
-    resolveId(importee: string, importer?: string) {
-      if (!importer || importee === entryPath) return importee
-      if (!importee[0].startsWith('.')) return null
-
-      const resolved = path.resolve(path.dirname(importer), importee)
-
-      if (resolved in absoluteFiles) return resolved
-      for (const ext of ['.js', '.tsx', '.jsx', '.ts']) {
-        const resolvedWithExt = `${resolved}${ext}`
-        if (resolvedWithExt in absoluteFiles) return resolvedWithExt
-      }
-      throw new Error(
-        `Could not resolve '${importee}' from ${
-          importer === entryPath
-            ? 'the entry MDX file'
-            : `'${importer.replace(`${dir}/`, '')}'`
-        }`,
+
+  const inMemoryPlugin: Plugin = {
+    name: 'inMemory',
+    setup(build) {
+      build.onResolve({filter: /.*/}, ({path: filePath, importer}) => {
+        if (filePath === entryPath) return {path: filePath}
+
+        const modulePath = path.resolve(path.dirname(importer), filePath)
+
+        if (modulePath in absoluteFiles) return {path: modulePath}
+
+        for (const ext of ['.js', '.ts', '.jsx', '.tsx', '.json', '.mdx']) {
+          const fullModulePath = `${modulePath}${ext}`
+          if (fullModulePath in absoluteFiles) return {path: fullModulePath}
+        }
+
+        return {
+          errors: [
+            {
+              text: `Could not resolve "${filePath}" in ${
+                importer === entryPath
+                  ? 'the entry MDX file.'
+                  : `"${importer.replace(dir, '.')}"`
+              }`,
+              location: null,
+            },
+          ],
+        }
+      })
+
+      build.onLoad(
+        {filter: /__mdx_bundler_fake_dir__/},
+        async ({path: filePath}) => {
+          const fileType = path.extname(filePath).slice(1)
+          const contents = absoluteFiles[filePath]
+
+          switch (fileType) {
+            case 'mdx': {
+              const vfile = await compileMDX(
+                {path: filePath, contents},
+                {
+                  remarkPlugins: [
+                    remarkFrontmatter,
+                    [remarkMdxFrontmatter, {name: 'frontmatter'}],
+                  ],
+                },
+              )
+              return {contents: vfile.contents.toString(), loader: 'jsx'}
```

Comment:
I think JSX is already compiled away by xdm unless you’re doing `jsx: true`, which I’m not seeing.

So, `loader: js` maybe?

Reply:
You could set `jsx: false` as well, so that xdm doesn’t turn jsx -> js. Then esbuild is doing all the jsx -> js transforms.

Reply:
It makes sense to use `jsx: false` I think then as you say esbuild is doing all the JSX work.